File: _ETC/B형대비문제/Social media/Social media/정답코드/Java/mid/2.java
//95,120KB	341MS	JAVA	4,564B
class UserSolution {
     
    private static int n;
    private static User[] user;
    private static Post[] post;
     
    private static Post[] feed;
     
    private static final int MAX = 100001;
     
    public void init(int N) {
//System.out.println("init: N=" + N);
 
        n = N + 1;
         
        user = new User[n];
        for (int i = 0 ; i < n ; i++) {
            user[i] = new User();
        }
 
        post = new Post[MAX];
        feed = new Post[11];
    }
 
    public void follow(int uID1, int uID2, int timestamp) {
//System.out.println("follow: uID1=" + uID1 + ", uID2=" + uID2 + ", timestamp=" + timestamp);
        user[uID1].follow[user[uID1].followcount++] = uID2;
    }
 
    public void makePost(int uID, int pID, int timestamp) {
//System.out.println("makePost: uID=" + uID + ", pID=" + pID + ", timestamp=" + timestamp);
        if (post[pID] == null) {
            post[pID] = new Post();
        }
        post[pID].pID = pID;
        post[pID].uID = uID;
        post[pID].timestamp = timestamp;
         
         
        user[uID].post[user[uID].postcount++] = pID;
    }
 
    public void like(int pID, int timestamp) {
//System.out.println("like: pID=" + pID + ", timestamp=" + timestamp);
        post[pID].like++;
    }
 
    public void getFeed(int uID, int timestamp, int pIDList[]) {
//System.out.println("getFeed: uID=" + uID + ", timestamp=" + timestamp);
         
//System.out.println("user["+uID+"].followcount=" + user[uID].followcount);
        for (int i = 0 ; i < user[uID].followcount ; i++) {
//System.out.println("user["+uID+"].follow["+i+"]=" + user[uID].follow[i]);
//System.out.println("user["+user[uID].follow[i]+"].postcount=" + user[user[uID].follow[i]].postcount);
        }
         
        int count = user[uID].postcount;
         
        for (int i = 0 ; i < user[uID].followcount ; i++) {
            count = count + user[user[uID].follow[i]].postcount;
        }       
//System.out.print("count=" + count + ", ");
 
        if (count > (10 * (user[uID].followcount + 1))) {
            count = (10 * (user[uID].followcount + 1));
        }
//System.out.print("(10 * (user["+uID+"].followcount + 1))=" + (10 * (user[uID].followcount + 1)));
         
        feed = new Post[count];
         
        int j = 0;
        for (int k = user[uID].postcount - 1, l = 0 ; k >= 0 && l < 10 ; k--, l++) {
            if (post[user[uID].post[k]].timestamp >= timestamp - 1000) {
                feed[j] = post[user[uID].post[k]];
                sort1(j++);
            } else {
                break;
            }
//System.out.println("uID=" + feed[j].uID + ", pID=" + feed[j].pID + ", like=" + feed[j].like + ", timestamp=" + feed[j].timestamp);
        }
         
        for (int i = 0 ; i < user[uID].followcount ; i++) {
            for (int k = user[user[uID].follow[i]].postcount - 1, l = 0 ; k >= 0 && l < 10 ; k--, l++) {
                if (post[user[user[uID].follow[i]].post[k]].timestamp >= timestamp - 1000) {
                    feed[j] = post[user[user[uID].follow[i]].post[k]];
                    sort1(j++);
                } else {
                    break;
                }
//System.out.println("uID=" + feed[j].uID + ", pID=" + feed[j].pID + ", like=" + feed[j].like + ", timestamp=" + feed[j].timestamp);
            }
        }
 
        for (int k = user[uID].postcount - 1, l = 0 ; k >= 0 && l < 10 ; k--, l++) {
            if (post[user[uID].post[k]].timestamp < timestamp - 1000) {
                feed[j] = post[user[uID].post[k]];
                sort2(j++);
            }
//System.out.println("uID=" + feed[j].uID + ", pID=" + feed[j].pID + ", like=" + feed[j].like + ", timestamp=" + feed[j].timestamp);
        }
         
        for (int i = 0 ; i < user[uID].followcount ; i++) {
            for (int k = user[user[uID].follow[i]].postcount - 1, l = 0 ; k >= 0 && l < 10 ; k--, l++) {
                if (post[user[user[uID].follow[i]].post[k]].timestamp < timestamp - 1000) {
                    feed[j] = post[user[user[uID].follow[i]].post[k]];
                    sort2(j++);
                }
//System.out.println("uID=" + feed[j].uID + ", pID=" + feed[j].pID + ", like=" + feed[j].like + ", timestamp=" + feed[j].timestamp);
            }
        }
         
        for (int k = 0 ; k < j && k < 10 ; k++) {
//System.out.println("uID=" + feed[k].uID + ", pID=" + feed[k].pID + ", like=" + feed[k].like + ", timestamp=" + feed[k].timestamp);
            pIDList[k] = feed[k].pID ;
        }
         
         
    }
     
    public static void sort1(int i) {
         
        Post temp = feed[i];
        int j = i - 1;
 
        while ((j >= 0) && ((temp.like > feed[j].like) || (temp.like == feed[j].like) && temp.timestamp > feed[j].timestamp)) {
            feed[j + 1] = feed[j];
            j--;
        }
        feed[j+1] = temp;
 
    }
 
    public static void sort2(int i) {
         
        Post temp = feed[i];
        int j = i - 1;
 
        while ((j >= 0) && temp.timestamp > feed[j].timestamp) {
            feed[j + 1] = feed[j];
            j--;
        }
        feed[j+1] = temp;
 
    }
 
    class User {
        int uID;
        int[] follow;
        int followcount;
        int[] post;
        int postcount;
        public User() {
            follow = new int[n];
            post = new int[MAX/500];
        }
    }
 
    class Post {
        int pID;
        int uID;
        int timestamp;
        int like;
    }
     
}

File: _ETC/B형대비문제/Social media/Social media/정답코드/CPP/mid/2.cpp
//22,752KB	151MS	C++	2,724B
#define MAX_USERS       1000
#define MAX_POSTINGS    100000
#define NULL 0
#define MAX_SIZE 11
 
typedef struct _post {
    int pid;
    int uid;
    int createdTime;
    int likes;
    _post* next;
} Post;
 
typedef struct _user {
    int nFollows;
    int following[MAX_USERS];
    int nPosts;
    Post* posts;
} User;
 
User users[MAX_USERS + 1];
Post posts[MAX_POSTINGS + 1];
 
int CURRENT_TIME;
bool operator<(Post a, Post b);
 
void init(int N)
{
    register int i;
    for (i = 1; i <= N; i++) {
        users[i].nFollows = 1;
        users[i].following[0] = i;
        users[i].nPosts = 0;
        users[i].posts = NULL;
    }
}
 
void follow(int uID1, int uID2, int timestamp)
{
    users[uID1].following[users[uID1].nFollows++] = uID2;
}
 
void makePost(int uID, int pID, int timestamp)
{
    posts[pID].pid = pID;
    posts[pID].uid = uID;
    posts[pID].createdTime = timestamp;
    posts[pID].likes = 0;
    if (users[uID].posts != NULL)
    {
        posts[pID].next = users[uID].posts;
    }
    else {
        posts[pID].next = NULL;
    }
    users[uID].posts = &posts[pID];
    users[uID].nPosts += 1;
}
 
void like(int pID, int timestamp)
{
    posts[pID].likes += 1;
}
 
void getFeed(int uID, int timestamp, int pIDList[])
{
    CURRENT_TIME = timestamp;
    Post heap[MAX_SIZE];
    register int i, j;
    int count = 0;
    for (i = 0; i < users[uID].nFollows; i++) {
        int u = users[uID].following[i];
 
        Post* cur = users[u].posts;
        for (j = 0; j < users[u].nPosts; j++) {
 
            if (count < 10) {
                heap[count].createdTime = cur->createdTime;
                heap[count].likes = cur->likes;
                heap[count].pid = cur->pid;
                heap[count].uid = cur->uid;
                count++;
                if (count >= 2) {
                    for (int m = count - 1; m > 0; m--) {
                        if (heap[m - 1] < heap[m]) {
                            // swap
                            Post tmp = heap[m];
                            heap[m] = heap[m - 1];
                            heap[m - 1] = tmp;
                        }
                        else {
                            break;
                        }
                    }
                }
            }
            else {
                if (heap[9] < *cur) {
                    heap[9].createdTime = cur->createdTime;
                    heap[9].likes = cur->likes;
                    heap[9].pid = cur->pid;
                    heap[9].uid = cur->uid;
 
                    for (int m = 9; m > 0; m--) {
                        if (heap[m - 1] < heap[m]) {
                            // swap
                            Post tmp = heap[m];
                            heap[m] = heap[m - 1];
                            heap[m - 1] = tmp;
                        }
                        else {
                            break;
                        }
                    }
                }
                else {
                    if ((cur->createdTime + 1000) <= CURRENT_TIME) {
                        break;
                    }
                }
            }
 
            cur = cur->next;
        }
    }
 
    for (i = 0; i < count; i++) {
        pIDList[i] = heap[i].pid;
    }
}
 
bool operator<(Post left, Post right)
{
    int val1, val2;
    if (left.createdTime + 1000 < CURRENT_TIME)
        val1 = left.createdTime;
    else
        val1 = left.likes * 100000 + left.createdTime;
 
    if (right.createdTime + 1000 < CURRENT_TIME)
        val2 = right.createdTime;
    else
        val2 = right.likes * 100000 + right.createdTime;
 
    return val1 < val2;
}